JSON 문자열 분석기 1단계 - 단순 List 분석

popsmile opened the pull request:
- 아래의 흐름으로 JSONParser를 구현했습니다.
  1. `InputView` 로 입력값을 받아온다.
  2.  입력받은 문자열을 `JSONGenerator` 을 통해 문자열 배열로 분리한다. 문자열을 다루는 기능을 `Extensions` 에 추가로 구현한다. 지원하지 않는 형식일 경우 문자열 배열 대신 `nil` 을 리턴하고, 안내메세지와 함께 프로그램을 종료한다.
  3. 리턴받은 문자열 배열에서 `TypeInspector` 로 특정 타입을 카운트한 결과를 `OutputView` 에서 출력한다.
- 설계한 기능들에 대해 각각 단위테스트를 시행했습니다.

popsmile committed:
Create InputView structure

popsmile committed:
Add String extensions

popsmile committed:
Create stringArray closure in JSONGenerator struct

Closure named 'stringArray' uses String extensions, which were added right before, to convert a String to String Array.

popsmile committed:
Create TypeInspector protocol

Add StringInspector, DoubleInspector, and BooleanInspector structures, conforming to the TypeInspector protocol. The protocol has countType() function to count certain type in a String array.

popsmile committed:
Create JsonParser struct in main to run this project

popsmile committed:
Add a print function in OutputView

Add showTypeCountOf() to print the result of parsing JSON data

popsmile committed:
Add a function not to print zero count of type

popsmile committed:
Add a function to check validity of input text as a JSON form

popsmile committed:
Create Unit Tests for JSONParser

popsmile committed:
Add some unit test functions of each file

popsmile committed:
Rename OutputView.Message.invalidJSON to invalidForm

godrm commented on JSONParser/JSONParserUnitTest/UnitTestTypeInspector.swift:
```diff
@@ -0,0 +1,38 @@
+class UnitTestTypeInspector: XCTestCase {
+    let jsonStringArray = ["10", "\"jk\"", "4", "\"314\"", "99", "\"crong\"", "false", "a"]
+    let jsonIncludingDouble = ["\"314\"", "99.12", "\"crong\"", "false"]
+    let jsonIncludeInvalidBool = ["flase", "ttrue", "Truee", "False"]
```
테스트 스위트에서 사용하는 더미 데이터가 모든 테스트함수에서 사용하나요?
그런게 아니면 필요한 함수 로컬에서만 선언해서 쓰세요.
참고로 모든 함수에서 필요한 객체가 있다면 setUp()에서 초기화 하면 됩니다.

popsmile replied:
넵. 참고하겠습니다 :)

popsmile replied:
넵. 참고하겠습니다 :)

godrm commented on JSONParser/JSONParser/OutputView.swift:
```diff
@@ -0,0 +1,58 @@
+        var typeCount = [Int]()
+        typeCount.append(StringInspector.countType(in: stringArray))
+        typeCount.append(DoubleInspector.countType(in: stringArray))
+        typeCount.append(BooleanInspector.countType(in: stringArray))
```
좋은 시도였네요.
하지만 이 코드 방식에서는 stringArray 배열이 길수록 처리 시간이 3배씩 느려지게 됩니다.
왜냐하면 모든 요소를 타입별로 반복해서 카운트를 하게 되니까요. 
이 부분이 아쉽네요.

popsmile replied:
앗 넵 이 점은 생각을 못했네요. ㅠㅠ 피드백 참고해서 다른 로직으로 한 번만 배열 요소를 탐색하도록 수정했습니다!

godrm commented on JSONParser/JSONParser/JSONGenerator.swift:
```diff
@@ -0,0 +1,18 @@
+import Foundation
+
+struct JSONGenerator {
+    private static let stringArray = { (jsonString: String) -> [String] in return jsonString.removeWhiteSpaces().trimSquareBrackets().splitByComma() }
```
JSONGenerator 는 JSON을 생성하는 객체인가요? 그런 의도가 맞나요?

popsmile replied:
그런 의도로 작성했는데, 막상 내부 기능은 문자열배열만 생성하고 있었습니다. 아래 피드백까지 참고해서 각각의 타입값을 그대로 저장하는 [Any] 배열을 생성하도록 추가했습니다. 이후에 다른 JSON 규격이 있다면 이 구조체에서 확장하고자 합니다.

godrm replied:
다시 한 번 생각해보세요. 지금은 JSON 데이터를 만드는게 아니라 문자열로 들어온 JSON 데이터를 스위프트 데이터 구조로 바꾸는 걸 만드는 겁니다 :)

godrm commented on JSONParser/JSONParser/JSONGenerator.swift:
```diff
@@ -0,0 +1,18 @@
+    
+    static func extractStringArray(from jsonString: String) -> [String]? {
+        guard jsonString.hasSideSquareBrackets() else { return nil }
+        return stringArray(jsonString)
```
이 미션에서 핵심은 입력받은 문자열을 잘라내서 그대로 `[String]` 문자열 배열로 만드는게 목표가 아닙니다.
각 타입별로 문자열이면 String, 숫자면 Int, 부울린이면 Boolean 타입 값이 배열에 들어가야합니다! 
이 부분을 고민해보세요!!

popsmile replied:
문자열 배열을 생성해서 처리하는 방식에서 더 나아가 각각의 타입값이 그대로 배열로 들어가도록 추가했습니다.

godrm commented on JSONParser/JSONParserUnitTest/UnitTestExtensions.swift:
```diff
@@ -0,0 +1,52 @@
+    let trimmedSquareBrackets = " 10, \"jk\", 4, \"314\", 99, \"crong\", false "
+    let stringSurroundedDoubleQuotation = "\"jamie\""
+    let trimmedDoubleQuotation = "jamie"
+    let splitArray = ["[ 10", " \"jk\"", " 4", " \"314\"", " 99", " \"crong\"", " false ]"]
```
테스트 스위트에서 사용하는 더미 데이터 중에서 상수더라도 테스트 함수 전체가 사용하는 게 아니라면
각 테스트 함수 내부에 선언해서 사용하세요.

popsmile committed:
Fix elements of a Set named boolExpression

popsmile committed:
Move the constant declarations of unit tests into each function

popsmile committed:
Add typeCast() method to convert [String] to [Any?] in JSONGenerator

popsmile committed:
Add struct TypeInspector to make a type count dictionary

Cut the protocol TypeInspector and the subprotocols, because they make the program runtime longer when the length of an array gets longer.

popsmile committed:
Create an enum typeName used to add a key to the dictionary typeCount

Move the struct Message into the struct OutputView and make it private.

popsmile committed:
Revise unit tests to conform to the changes of each file

godrm commented:
1단계 경험은 충분한 것 같습니다. 
남은건 다음 단계로 넘어가서 함께 개선해보세요. 
한 가지 덧붙여서, Any를 쓰기 보다 좀더 구체적으로 JSON에서만 쓰는 타입들을 포함하는 타입을 지정해보세요!

popsmile commented:
네 :) 알겠습니다!
